django/db/models/fields/related_descriptors.py

```diff
@@ -1092,16 +1102,42 @@ def _apply_rel_filters(self, queryset):
             return queryset._next_is_sticky().filter(**self.core_filters)
 
         def get_prefetch_cache(self):
-            try:
-                return self.instance._prefetched_objects_cache[self.prefetch_cache_name]
-            except (AttributeError, KeyError):
-                return None
+            current_instance = self.instance
+            while True:
+                try:
+                    return current_instance._prefetched_objects_cache[
+                        self.prefetch_cache_name
+                    ]
+                except (AttributeError, KeyError):
+                    pass
+                # Walk up the ancestor-chain (if cached) to try and find
+                # a prefetch in an ancestor
+                ancestor_link = current_instance._meta.get_ancestor_link(
+                    rel.field.model
+                )
+                if ancestor_link and ancestor_link.is_cached(current_instance):
+                    current_instance = ancestor_link.get_cached_value(current_instance)
```

Comment:
I think we could avoid the double function call by systematically calling `get_cached_value`

```suggestion
                if not (ancestor_link and current_instance := ancestor_link.get_cached_value(current_instance)):
                    return None
```